### PR TITLE
feat: add Scam Sniffer [ethshanghai]

### LINKED ISSUE #6311: [Bounty] Enhance your Web3 experience with Mask
## Catalogue

- Platform
  - [x] Gitcoin
- Experience
  - [x] Intermediate
- Category
  - [x] Mask Plugin
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] $5000

**ETH SHANGHAI ONLY**

## Bounty Description

Mask has been working on bringing the latest technologies and best features to improve user experience while browsing major social media platforms, such as Twitter. We've already enhanced the Twitter timeline, including enabling a comprehensive trading view floating window while hovering your mouse on any cashtag and rendering an interface for any opensea links showing the corresponding NFT and related listing info, as shown below. We have also included many community plugins to better empower the timeline.

image
image

Now, please turn your eyes into your profile page. We've also injected a `Web3` tab showing your `Web3` connections and assets. You can connect your `persona` with your blockchain wallets and your social media accounts via  `Next.ID`. All the related assets like your NFTs, your donations on gitcoin or your poaps are listed under this `Web3` tab, thanks to `RSS3`. You could also set your favorite nft as your `non-fungible friend` to welcome friends that are viewing your profile page.

image

image

This time in this hackathon, we would like to take this step even further: how can you possibly enhance your Web3 experience into another level? Social games, social profile building, or anything that can help social media users to maximize their social experience. 

Open your mind and build it for us. We will choose up to two winners in this hackathon and each winner would be able to win $2500 or more!

## Winner Info

We will select up to **2** winners and each will receive at least $2500 as the winning prize.

## Submission Info

Please make sure you read through our [docs](https://github.com/DimensionDev/Maskbook/tree/develop/docs), especially the [bounty guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/bounty-development-guide.md) before starting your hacking. You will need to **create a pr** to this repo to submit your work.

## Contact Info

@guanbinrui @Jack-Works for any question regarding how to work on Mask.

### PR DESCRIPTION
**Pull request by lljxx1**
## Description
A plugin which can help people against with the crypto scammers

**Website**: https://scamsniffer.io/

#### How it's work:
- Aggregate popular projects from Coingecko and Opensea
- Once detect the projects mentioned in tweet, links and twitter account information will be checked
- Show scam alert

Closes #6311

### Demo Video
https://www.youtube.com/watch?v=s9BXAQb0SjU

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

##### Scam Tweets
https://twitter.com/NFT_Adidas/status/1528523129160597510
https://twitter.com/apecoinofficiaI/status/1521570752138342400

image

image

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/plugins/ScamSniffer/src/base.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { languages } from './locales/languages'
+
+export const base: Plugin.Shared.Definition = {
+    ID: PLUGIN_ID,
+    name: { fallback: PLUGIN_NAME },
+    description: { fallback: PLUGIN_DESCRIPTION },
+    publisher: { name: { fallback: 'Scam Sniffer' }, link: 'https://scamsniffer.io/' },
+    enableRequirement: {
+        architecture: { app: false, web: true },
+        networks: { type: 'opt-out', networks: {} },
```
Comment (Jack-Works) [inline packages/plugins/ScamSniffer/src/base.ts:12]: Does it work for all networks? if not, please specify the network you supports.

Comment (lljxx1) [reply packages/plugins/ScamSniffer/src/base.ts:12]: yeah, It's not depends on any network

File: packages/plugins/ScamSniffer/src/SNSAdaptor/ReportButton.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            slug: 'report_button',
+            name: 'report_button',
+            matchType: 'report_button',
+            externalUrl: null,
+            twitterUsername: null,
+            post: postDetail,
+        })
+        await new Promise((resolve) => {
+            setTimeout(resolve, 1200)
+        })
```
Comment (guanbinrui) [inline packages/plugins/ScamSniffer/src/SNSAdaptor/ReportButton.tsx]:
```suggestion
        await delay(1200)
```

File: packages/plugins/ScamSniffer/src/SNSAdaptor/ScamAlert.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+const ScamAlert = ({ result }: { result: ScamResult }) => {
+    const { classes } = useStyles()
+    const [autoReport, setAutoReport] = useState(false)
+    const t = useI18N()
+
+    useEffect(() => {
+        if (autoReport) {
+            PluginScamRPC.sendReportScam(result)
+        }
+    }, [autoReport])
```
Comment (guanbinrui) [inline packages/plugins/ScamSniffer/src/SNSAdaptor/ScamAlert.tsx]:
```suggestion
    }, [autoReport, result])
```

**Comment by lljxx1** [comment]
@guanbinrui ci build error: Allocation failed - JavaScript heap out of memory

**Comment by guanbinrui** [comment]
> @guanbinrui ci build error: Allocation failed - JavaScript heap out of memory

It could cause by the modification in package-lcok.json. So please copy it from the develop branch, and reinstall modules again.

Learn more: https://docs.mask.io/docs/faq#how-to-resolve-merge-conflicts-in-pnpm-lockyaml

**Comment by lljxx1** [comment]
> > @guanbinrui ci build error: Allocation failed - JavaScript heap out of memory
> 
> It could cause by the modification in package-lcok.json. So please copy it from the develop branch, and reinstall modules again.
> 
> Learn more: https://docs.mask.io/docs/faq#how-to-resolve-merge-conflicts-in-pnpm-lockyaml

yeah, I did that, but still have the same problem

**Review by guanbinrui — DISMISSED** [review]

**Comment by AinuX** [comment]
@lljxx1 
With latest commit build (https://github.com/DimensionDev/Maskbook/commit/c7be0b398990c55bb66b93fffd2a49b5a3bc8aa5), seems that it did not work.

Open scam Tweet(https://twitter.com/apecoinofficiaI/status/1521570752138342400), but there is no phishing sniffer plugin was rendered, please check, thanks!

**Comment by lljxx1** [comment]
> @lljxx1 With latest commit build ([c7be0b3](https://github.com/DimensionDev/Maskbook/commit/c7be0b398990c55bb66b93fffd2a49b5a3bc8aa5)), seems that it did not work.
> 
> Open scam Tweet(https://twitter.com/apecoinofficiaI/status/1521570752138342400), but there is no phishing sniffer plugin was rendered, please check, thanks!

it's works on my browser:
image

try these tweets: 
https://twitter.com/animebr56014480/status/1540016514778136577
https://twitter.com/savanth_harsha/status/1539961459265765376

**Comment by lljxx1** [comment]
> > @lljxx1 With latest commit build ([c7be0b3](https://github.com/DimensionDev/Maskbook/commit/c7be0b398990c55bb66b93fffd2a49b5a3bc8aa5)), seems that it did not work.
> > Open scam Tweet(https://twitter.com/apecoinofficiaI/status/1521570752138342400), but there is no phishing sniffer plugin was rendered, please check, thanks!
> 
> it's works on my browser: image
> 
> try these tweets: https://twitter.com/animebr56014480/status/1540016514778136577 https://twitter.com/savanth_harsha/status/1539961459265765376

and fixed some icon size issue

**Pull request merged**